File: linkedlist/reverselinkedlist/cs/ii/ReverseLinkedList/ReverseLinkedList.cs
namespace ReverseLinkedListLib;

public class ListNode
{
    public int val { get; set; }
    public ListNode next { get; set; }

    public ListNode(int val)
    {
        this.val = val;
    }
}

public class ReverseLinkedList
{
    /*
     *
     *  1 -> 2 -> 3 -> 4 -> 5 -> null
     *  
     *  5 -> 4 -> 3 -> 2 -> 1 -> null
     *
     *
     *  1 -> 2 -> 3 -> null
     *
     *  2 -> 1
     *  
     *  3 -> 2
     *
     */
    public ListNode ReverseList(ListNode head)
    {
        if (head.next == null)
            return head;

        var res = ReverseList(head.next);
        head.next.next = head;
        head.next = null;
        return res;
    }
}

File: linkedlist/reverselinkedlist/py/reverselinkedlist.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if head == None or head.next == None: 
            return head

        r = self.reverseList(head.next)

        head.next.next = head
        head.next = None

        return r


sol = Solution()

head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)

r = sol.reverseList(head)

print(r)

